index.js

```diff
@@ -2,73 +2,119 @@
 const errors = require('restify-errors')
 const joi = require('joi')
 
-const defaultErrorTransformer = (validationInput, joiError) => {
-  const retError = new errors.BadRequestError()
-  retError.body.message = joiError.message
-  retError.body.data = joiError.details
-  return retError
+const middlewareDefaults = {
+  joiOptions: {
+    convert: true,
+    allowUnknown: false,
+    abortEarly: false
+  },
+  errorTransformer: (validationInput, joiError) => {
+    const retError = new errors.BadRequestError()
+    retError.body.message = joiError.message
+    retError.body.data = joiError.details
+    return retError
+  },
+  errorResponder: (transformedErr, req, res, next) => next(transformedErr),
+  keysToValidate: ['params', 'body', 'query', 'user', 'headers', 'trailers', 'files']
+}
+
+const optionsSchema = {
+  joiOptions: joi.object(),
+  errorTransformer: joi.func(),
+  errorResponder: joi.func(),
+  keysToValidate: joi.array().items(joi.string())
+}
+
+const checkOptions = (defaults, input) => {
+  const {joiOptions, errorTransformer, errorResponder, keysToValidate} = defaults
+  const schema = {
+    joiOptions: optionsSchema.joiOptions.default(joiOptions),
+    errorTransformer: optionsSchema.errorTransformer.default(errorTransformer),
+    errorResponder: optionsSchema.errorResponder.default(errorResponder),
+    keysToValidate: optionsSchema.keysToValidate.default(keysToValidate)
+  }
+
+  const {value, error} = joi.validate(input, schema, middlewareDefaults.joiOptions)
+  if (error) {
+    error.message = `Error: restify-joi-middleware, bad configuration found: ${error.message}`
+    throw error
+  }
+  return value
 }
 
-const defaultErrorResponder = (transformedErr, req, res, next) => next(transformedErr)
+// should this use a lru cache?
+const routeOptionsCache = {}
```

Comment:
Not sure about this yet, for APIs with lots of routes, this could eat a bit of memory, not sure how common that case is.